azuredevops/internal/service/memberentitlementmanagement/resource_service_principal_entitlement.go

```diff
@@ -0,0 +1,322 @@
+		ServicePrincipalId: &id,
+	})
+
+	if err != nil {
+		return fmt.Errorf("Deleting service principal entitlement: %v", err)
+	}
+
+	return nil
+}
+
+func resourceServicePrincipalEntitlementUpdate(d *schema.ResourceData, m interface{}) error {
+	servicePrincipalEntitlementID := d.Id()
+	id, err := uuid.Parse(servicePrincipalEntitlementID)
+	if err != nil {
+		return fmt.Errorf("Parsing ServicePrincipalEntitlement ID. ServicePrincipalEntitlementID: %s. %v", servicePrincipalEntitlementID, err)
+	}
+
+	accountLicenseType, err := converter.AccountLicenseType(d.Get("account_license_type").(string))
+	if err != nil {
+		return err
+	}
+	licensingSource, ok := d.GetOk("licensing_source")
+	if !ok {
+		return fmt.Errorf("Reading account licensing source for ServicePrincipalEntitlementID: %s", servicePrincipalEntitlementID)
+	}
+
+	clients := m.(*client.AggregatedClient)
+
+	patchResponse, err := clients.MemberEntitleManagementClient.UpdateServicePrincipalEntitlement(clients.Ctx,
+		memberentitlementmanagement.UpdateServicePrincipalEntitlementArgs{
+			ServicePrincipalId: &id,
+			Document: &[]webapi.JsonPatchOperation{
+				{
+					Op:   &webapi.OperationValues.Replace,
+					From: nil,
+					Path: converter.String("/accessLevel"),
+					Value: struct {
+						AccountLicenseType string `json:"accountLicenseType"`
+						LicensingSource    string `json:"licensingSource"`
+					}{
+						string(*accountLicenseType),
+						licensingSource.(string),
+					},
+				},
+			},
+		})
+
+	if err != nil {
+		return fmt.Errorf("Updating service principal entitlement: %v", err)
+	}
+
+	if !*patchResponse.IsSuccess {
+		return fmt.Errorf("Updating service principal entitlement: %s", getServicePrincipalEntitlementAPIErrorMessage(patchResponse.OperationResults))
+	}
+	return resourceServicePrincipalEntitlementRead(d, m)
+}
+
+func importServicePrincipalEntitlement(d *schema.ResourceData, m interface{}) ([]*schema.ResourceData, error) {
+	upn := d.Id()
+	id, err := uuid.Parse(upn)
+
+	if err != nil {
+		return nil, fmt.Errorf("Only UUID values can used for import [%s]", upn)
+	}
+
+	clients := m.(*client.AggregatedClient)
+	result, err := clients.MemberEntitleManagementClient.GetServicePrincipalEntitlement(clients.Ctx, memberentitlementmanagement.GetServicePrincipalEntitlementArgs{
+		ServicePrincipalId: &id,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("Error getting the service principal entitlement with supplied id %s: %s", upn, err)
+	}
+
+	d.SetId((*result).Id.String())
+
+	return []*schema.ResourceData{d}, nil
+}
+
+func flattenServicePrincipalEntitlement(d *schema.ResourceData, servicePrincipalEntitlement *memberentitlementmanagement.ServicePrincipalEntitlement) {
+	d.SetId(servicePrincipalEntitlement.Id.String())
+	d.Set("descriptor", *servicePrincipalEntitlement.ServicePrincipal.Descriptor)
```

Comment:
Should check the `servicePrincipalEntitlement.ServicePrincipal.Descriptor` and `servicePrincipalEntitlement.ServicePrincipal` first before get the values to prevent potential nil exception

Reply:
I'm not sure I understand the request. The check is done in l126, before the function is called. I can repeat the same check within the function and throw an error if that is the correct way to do it

Reply:
Service may not return the property due to permission. Therefore, you need to check this property before calling it with a pointer.